### PR TITLE
feat(insured-bridge): Implement settleRelay

### PR DESCRIPTION
**Pull request by nicholaspai**
**Motivation**

This method can be called after the Oracle has resolved a price for a relay price request. It pays out relayer rewards and the recipient's relay.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol [hunk 1]
```diff
@@ -211,15 +215,63 @@ contract BridgePool is Testable {
...
+            ),
+            "OptimisticOracle has not resolved relay price request"
+        );
+
+        // Note: Why don't we have to check the value of the price?
+        // - If the OptimisticOracle has a price and the relayState is PENDING, then we can safely assume that the relay
+        // was validated. This is because this contract proposes a price of 1e18, or "YES" to the identifier posing the
+        // question "Is this relay valid for the associated deposit?". If the proposal is disputed, then the relayState
+        // will be reset to UNINITIALIZED. If the proposal is not disputed, and there is a price available, then the
+        // proposal must have passed the dispute period.
```
Comment (chrismaree) [inline packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol:239]:
```suggestion
        // proposal must have passed the dispute period, assuming the proposal passed OO liveness.
```

File: packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol [hunk 2]
```diff
@@ -211,15 +215,63 @@ contract BridgePool is Testable {
...
-     * be one instant relayer per relay attempt and disputed relays cannot be sped up.
-     * @param _depositData Unique set of L2 deposit data that caller is trying to instantly relay.
+     * @param _depositData Unique set of L2 deposit data that caller is trying to settle a relay for.
      */
     function settleRelay(DepositData memory _depositData) public {
         bytes32 depositHash = _getDepositHash(_depositData);
         RelayData storage relay = relays[depositHash];
         require(relays[depositHash].relayState == RelayState.Pending, "Can only speed up pending slow relay");
-        // TODO: Pay relayer rewards using PendingRelay data.
+        require(
```
Comment (chrismaree) [inline packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol:224]: prehaps add a note that this method will return false if liveness has not passed for the price request. this is key in the context of the bridge as this implies that the action is "valid".

File: packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol [hunk 3]
```diff
@@ -211,15 +215,63 @@ contract BridgePool is Testable {
...
      * the relay as complete.
-     * @dev Caller must have approved this contract to spend the deposit amount of L1 tokens to relay. There can only
-     * be one instant relayer per relay attempt and disputed relays cannot be sped up.
-     * @param _depositData Unique set of L2 deposit data that caller is trying to instantly relay.
+     * @param _depositData Unique set of L2 deposit data that caller is trying to settle a relay for.
      */
     function settleRelay(DepositData memory _depositData) public {
         bytes32 depositHash = _getDepositHash(_depositData);
         RelayData storage relay = relays[depositHash];
         require(relays[depositHash].relayState == RelayState.Pending, "Can only speed up pending slow relay");
```
Comment (chrismaree) [inline packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol:223]: Is this require message not wrong? should it not relate to the validity of the relay action?

File: packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol [hunk 4]
```diff
@@ -211,15 +215,63 @@ contract BridgePool is Testable {
...
+                getRelayAncillaryData(_depositData, relay)
+            ),
+            "OptimisticOracle has not resolved relay price request"
+        );
+
+        // Note: Why don't we have to check the value of the price?
+        // - If the OptimisticOracle has a price and the relayState is PENDING, then we can safely assume that the relay
+        // was validated. This is because this contract proposes a price of 1e18, or "YES" to the identifier posing the
+        // question "Is this relay valid for the associated deposit?". If the proposal is disputed, then the relayState
+        // will be reset to UNINITIALIZED. If the proposal is not disputed, and there is a price available, then the
```
Comment (chrismaree) [inline packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol:238]:
```suggestion
        // will be reset to UNINITIALIZED from the OO callback. If the proposal is not disputed, and there is a price available, then the
```

File: packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol [hunk 5]
```diff
@@ -211,15 +215,63 @@ contract BridgePool is Testable {
...
+        relay.relayState = RelayState.Finalized;
+
+        // Pay recipient the deposit amount less fees.
+        IERC20(_depositData.l1Token).safeTransfer(
+            _depositData.recipient,
+            _depositData.amount - _getRealizedFeeAmount(relay.realizedFeePct, _depositData.amount)
+        );
+
+        // Reward relayers.
+        // TODO: For now, if there was an instant relay associated with this deposit, then split the reward 50/50
```
Comment (chrismaree) [inline packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol:250]: this seems fine for now. will refactor this to consider all fees defined by the depositor in a later PR once all LP logic is done + e2e tests for this.

File: packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol [hunk 6]
```diff
@@ -211,15 +215,63 @@ contract BridgePool is Testable {
...
+        // between slow and instant relayer. Otherwise pay the slow relayer the full reward.
+        uint256 rewardPool = _getProposerRewardAmount(relay.proposerRewardPct, _depositData.amount);
+        if (relay.instantRelayer != address(0)) {
```
Comment (chrismaree) [inline packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol:253]: this bit of branching logic is really important as it defines the kind of payout done. please add another comment on this.

**Review by chrismaree — APPROVED** [review]
Nice simple code. good work.

At this point we really need to start thinking about all the edge cases that might come from the pattern we've chosen. I'll try work through some of these in e2e tests as well.
